src/librustc_resolve/resolve_imports.rs

```diff
@@ -142,110 +135,84 @@ impl<'a, 'crateloader> Resolver<'a, 'crateloader> {
-                    // if a crate with the same name is found in `extern_prelude`,
-                    // preventing the check below this one from returning `binding`
-                    // in all cases.
-                    //
-                    // That is, if there's no crate with the same name, `binding`
-                    // is always returned, which is the result of doing the exact
-                    // same lookup of `ident`, in the `self` module.
-                    // But when a crate does exist, it will get chosen even when
-                    // macro expansion could result in a success from the lookup
-                    // in the `self` module, later on.
-                    //
-                    // NB. This is currently alleviated by the "ambiguity canaries"
-                    // (see `is_uniform_paths_canary`) that get introduced for the
-                    // maybe-relative imports handled here: if the false negative
-                    // case were to arise, it would *also* cause an ambiguity error.
-                    if binding.is_ok() {
-                        return binding;
+            ModuleOrUniformRoot::UniformRoot(uniform_root_kind) => {
+                assert!(!restricted_shadowing);
+                match uniform_root_kind {
+                    UniformRootKind::ExternPrelude => {
+                        return if let Some(binding) =
+                                self.extern_prelude_get(ident, !record_used, false) {
+                            Ok(binding)
+                        } else if !self.graph_root.unresolved_invocations.borrow().is_empty() {
+                            // Macro-expanded `extern crate` items can add names to extern prelude.
+                            Err(DeterminacyExt::Undetermined)
+                        } else {
+                            Err(DeterminacyExt::Determined)
+                        }
                     }
+                    UniformRootKind::CurrentScope => {
+                        let parent_scope =
+                            parent_scope.expect("no parent scope for a single-segment import");
+
+                        if ns == TypeNS {
+                            if ident.name == keywords::Crate.name() ||
+                               ident.name == keywords::DollarCrate.name() {
+                                let module = self.resolve_crate_root(ident);
+                                let binding = (module, ty::Visibility::Public,
+                                               module.span, Mark::root())
+                                               .to_name_binding(self.arenas);
+                                return Ok(binding);
+                            } else if ident.name == keywords::Super.name() ||
+                                      ident.name == keywords::SelfValue.name() {
+                                // FIXME: Implement these with renaming requirements so that e.g.
```

Comment:
Is this preexisting? Should this be `bug!` or any other kind of failure so that it does not get silently forgotten?

Reply:
Yes, preexisting.
This is not a bug, this happens e.g. when someone writes `use super as name;`.
Until this is properly implemented we simply fall through into `early_resolve_ident_in_lexical_scope` and get `Err(Determined)`.